placement/algo/mirrored.go

```diff
@@ -285,6 +267,83 @@ func allInstancesInState(
 
 	return len(instanceIDs) == 0
 }
+
+func (a mirroredAlgorithm) returnInitializingShards(
+	p placement.Placement,
+	instanceIDs []string,
+) (placement.Placement, error) {
+	shouldContinue := true
+	for shouldContinue {
+		shouldContinue = false
+		for _, id := range instanceIDs {
+			_, exist := p.Instance(id)
+			if !exist {
+				continue
+			}
+			ph, instance, err := newRemoveInstanceHelper(p, id, a.opts)
+			if err != nil {
+				return nil, err
+			}
+			numInitShards := instance.Shards().NumShardsForState(shard.Initializing)
+			ph.ReturnInitializingShards(instance)
+			if instance.Shards().NumShardsForState(shard.Initializing) < numInitShards {
+				// Made some progress on returning shards.
+				shouldContinue = true
+			}
+			p = ph.GeneratePlacement()
+			if instance.Shards().NumShards() > 0 {
+				p = p.SetInstances(append(p.Instances(), instance))
+			}
+		}
+	}
+
+	for _, id := range instanceIDs {
+		instance, ok := p.Instance(id)
+		if !ok {
+			continue
+		}
+		numInitializingShards := instance.Shards().NumShardsForState(shard.Initializing)
+		if numInitializingShards != 0 {
+			return nil, fmt.Errorf("there are %d initializing shards could not be returned for instance %s", numInitializingShards, id)
+		}
+	}
+
+	return p, nil
+}
+
+func (a mirroredAlgorithm) reclaimLeavingShards(
+	p placement.Placement,
+	addingInstances []placement.Instance,
+) (placement.Placement, error) {
+	shouldContinue := true
+	for shouldContinue {
+		shouldContinue = false
+		for _, instance := range addingInstances {
+			ph, instance := newAddInstanceHelper(p, instance, a.opts)
+			numLeavingShards := instance.Shards().NumShardsForState(shard.Leaving)
+			ph.ReclaimLeavingShards(instance)
+			if instance.Shards().NumShardsForState(shard.Leaving) < numLeavingShards {
+				// Made some progress on reclaiming shards.
+				shouldContinue = true
+			}
+			p = ph.GeneratePlacement()
+		}
+	}
+
+	for _, instance := range addingInstances {
+		instance, ok := p.Instance(instance.ID())
+		if !ok {
+			return nil, fmt.Errorf("could not find instance %s in placement after reclaiming leaving shards", instance.ID())
```

Comment:
Probably should do
```
instanceID := instance.ID()
instance, ok := p.Instance(InstanceID)
if !ok {
   return nil, fmt.Errorf("...", instanceID)
}
...
```
? Otherwise it might panic.

Reply:
yeah good catch